### PR TITLE
Translator

### PR DESCRIPTION
**Pull request by skad0**
@AnyaTyu  @tadatuta 
need code review

### CODE REVIEW

**Comment by AnyaTyu** [comment]
@skad0 squash all into one commit with name for example "add api ya.translator" and rename PR accordingly

File: common.blocks/editor/editor.js [hunk 1]
```diff
@@ -9,27 +9,26 @@ provide(BEMDOM.decl(this.name, {
             .toggleMod(target, 'focused')
             .toggleMod(source, 'focused');
     },
-    provideData: function() {
-        const items = this.elem('wrapper');
-        const _this = this;
+    provideData: function(options = {}) {
+        const items = this.domElem.serializeArray();
+        const onlyTranslated = options.onlyTranslated === undefined ? true : false;
```
Comment (AnyaTyu) [inline common.blocks/editor/editor.js]: Why so? !!undefined === false

File: common.blocks/editor/editor.js [hunk 2]
```diff
@@ -9,27 +9,26 @@ provide(BEMDOM.decl(this.name, {
...
 
         let data = [];
 
-        items.each((index, item) => {
-            const jqueryItem = $(item);
-            const source = _this.findElem(jqueryItem, 'source').text();
-            const target = _this.findBlockInside(jqueryItem, 'textarea').getVal();
-            const status = _this.findBlockInside(jqueryItem, 'checkbox').getMod('checked');
+        for (let i = 0; i < items.length; i += 3) {
```
Comment (AnyaTyu) [inline common.blocks/editor/editor.js:18]: please add comment for this segment of code

File: common.blocks/workspace/workspace.js [hunk 3]
```diff
@@ -17,18 +17,24 @@ modules.define('workspace', ['i-bem__dom', 'jquery', 'info-modal'],
...
-            $.get('/translation' + location.search)
+
+        _getTranslation: function() {
+            const data = this._editor.provideData({ onlyTranslated: false });
+            InfoModal.show('Loading');
+            $.post('/translate', { data: JSON.stringify(data) })
                 .then((response) => {
-                    console.log(response);
+                    InfoModal.hide();
+                    BEMDOM.replace(this._editor.domElem, response);
```
Comment (AnyaTyu) [inline common.blocks/workspace/workspace.js]: I think variable _editor need update for example: this._editor = this.findBlockInside('editor');

@tadatuta ?

File: package.json [hunk 4]
```diff
@@ -7,7 +7,7 @@
     "url": "git@github.com:bem-site/xliff-editor.git"
   },
   "engines": {
-    "node": ">=4"
+    "node": ">=6"
```
Comment (AnyaTyu) [inline package.json:10]: 👍

File: server/controllers/translator.js [hunk 5]
```diff
@@ -57,7 +58,50 @@ function saveMemory(req, res) {
 }
 
 function getTranslate(req, res) {
+    const passport = req.session.passport || {};
+    const token = passport.user && passport.user.token;
```
Comment (AnyaTyu) [inline server/controllers/translator.js]: token do not use

**Comment by tadatuta** [comment]
rebase needed

File: common.blocks/editor/editor.bemtree.js [hunk 6]
```diff
@@ -1,13 +1,29 @@
...
         return {
             elem: 'wrapper',
+            elemMods: { verified: status },
             content: [
                 {
                     elem: 'source',
-                    content: segment.source.content
+                    content: [
+                        {
+                            elem: 'input',
```
Comment (tadatuta) [inline common.blocks/editor/editor.bemtree.js]: any reason to make it an element?

File: common.blocks/editor/editor.bemtree.js [hunk 7]
```diff
@@ -1,13 +1,29 @@
 block('editor').content()((node) => {
     const segments = node.data.segments || [];
 
+    var counter = 0;
```
Comment (tadatuta) [inline common.blocks/editor/editor.bemtree.js]: Use second argument of `Array.prototype.map` instead.

File: common.blocks/editor/editor.bemtree.js [hunk 8]
```diff
@@ -16,14 +32,16 @@ block('editor').content()((node) => {
...
+                        val: segment.target.content || '',
+                        name: 'target_' + counter
                     }
                 },
                 {
                     block: 'checkbox',
                     mix: { block: 'editor', elem: 'status' },
-                    name: 'status',
-                    val: segment.status
+                    mods: { checked: (status === 'on' || status === true) },
```
Comment (tadatuta) [inline common.blocks/editor/editor.bemtree.js]: can it be shorten to `checked: status !== 'off'`?

File: common.blocks/editor/editor.bemtree.js [hunk 9]
```diff
@@ -16,14 +32,16 @@ block('editor').content()((node) => {
...
+                    name: 'status_' + counter++,
+                    val: 'on'
```
Comment (tadatuta) [inline common.blocks/editor/editor.bemtree.js:41]: is it really needed?

File: common.blocks/editor/editor.css [hunk 10]
```diff
@@ -49,3 +49,10 @@
           border-color: #079;
       }
 }
+
+.editor_unverified .editor__wrapper_verified {
+    display: block;
+    height: 2px;
+    background: red;
+    overflow: hidden;
+}
```
Comment (tadatuta) [inline common.blocks/editor/editor.css]: new line at EOF

File: common.blocks/editor/editor.js [hunk 11]
```diff
@@ -9,27 +9,26 @@ provide(BEMDOM.decl(this.name, {
...
             .toggleMod(source, 'focused');
     },
-    provideData: function() {
-        const items = this.elem('wrapper');
-        const _this = this;
+    provideData: function(options = {}) {
+        const items = this.domElem.serializeArray();
+        const onlyTranslated = options.onlyTranslated === undefined ? true : options.onlyTranslated;
 
         let data = [];
```
Comment (tadatuta) [inline common.blocks/editor/editor.js]: const

File: server/app.js [hunk 12]
```diff
@@ -18,8 +19,8 @@ const app = express();
 
 app
     // .use(favicon(path.join(path.join(rootDir, 'index'), 'favicon.ico')))
-    .use(bodyParser.json())
-    .use(bodyParser.urlencoded({ extended: false }))
+    .use(bodyParser.json({ limit: '1mb' }))
```
Comment (tadatuta) [inline server/app.js]: let's make it 3mb

File: server/controllers/translator.js [hunk 13]
```diff
@@ -52,7 +57,49 @@ function saveMemory(req, res) {
...
+            }
+            translate.translate(item.source, {
+                // TODO: here we should think for better solution
+                // maybe we should use simplified format anywhere
+                from: item.source_lang.slice(0,2),
+                to:   item.target_lang.slice(0,2)
+            },
+            function(err, result) {
+                if (err) {
+                    reject({ code: 500, message: 'Something wrong happens' });
```
Comment (tadatuta) [inline server/controllers/translator.js]: please add `console.error(err);` here

File: server/controllers/translator.js [hunk 14]
```diff
@@ -52,7 +57,49 @@ function saveMemory(req, res) {
...
+                }
 
+                if (result.code === 200) {
+                    item.target = { content: result.text[0] };
+                    item.source = { content: item.source };
+                    return resolve(item);
+                } else {
+                    reject(result);
```
Comment (tadatuta) [inline server/controllers/translator.js]: add logs

File: server/controllers/translator.js [hunk 15]
```diff
@@ -52,7 +57,49 @@ function saveMemory(req, res) {
...
+                }
+            });
+        });
+    }))
+        .then((data) => {
```
Comment (tadatuta) [inline server/controllers/translator.js]: no need in brackets if there's just one argument

File: common.blocks/editor/editor.bemtree.js (hunk 6 above)
Comment (skad0) [inline common.blocks/editor/editor.bemtree.js]: why not?

File: .gitignore [hunk 16]
```diff
@@ -6,6 +6,8 @@ node_modules/
 .svn
 npm-debug.log
 server/secret-config.js
+data/
+.env
```
Comment (AnyaTyu) [inline .gitignore:10]: secret-config --> env

File: common.blocks/editor/editor.bemtree.js [hunk 17]
```diff
@@ -1,7 +1,7 @@
 block('editor').content()((node) => {
     const segments = node.data.segments || [];
 
-    return segments.map(segment => {
+    return segments.map((segment, counter) => {
```
Comment (AnyaTyu) [inline common.blocks/editor/editor.bemtree.js]: idx

File: common.blocks/editor/editor.bemtree.js [hunk 18]
```diff
@@ -19,15 +30,16 @@ block('editor').content()((node) => {
...
                 },
                 {
                     block: 'checkbox',
                     mix: { block: 'editor', elem: 'status' },
-                    mods: { checked: status },
-                    name: 'status',
-                    val: status
+                    mods: { checked: (status && status !== 'off') },
+                    name: 'status_' + counter,
+                    val: 'on'
```
Comment (AnyaTyu) [inline common.blocks/editor/editor.bemtree.js:41]: ?

File: common.blocks/editor/editor.js [hunk 19]
```diff
@@ -9,27 +9,25 @@ provide(BEMDOM.decl(this.name, {
             .toggleMod(target, 'focused')
             .toggleMod(source, 'focused');
     },
-    provideData: function() {
-        const items = this.elem('wrapper');
-        const _this = this;
+    provideData: function(options = {}) {
```
Comment (AnyaTyu) [inline common.blocks/editor/editor.js:12]: getData

File: server/controllers/translator.js [hunk 20]
```diff
@@ -57,7 +57,51 @@ function saveMemory(req, res) {
 }
 
 function getTranslate(req, res) {
+    const passport = req.session.passport || {};
+    const data = JSON.parse(req.body.data);
+
+    return Promise.all(data.map((item) => {
```
Comment (AnyaTyu) [inline server/controllers/translator.js]: не нужны скобки

File: server/controllers/translator.js [hunk 21]
```diff
@@ -3,6 +3,7 @@ const md2xliff = require('md2xliff');
 const GitHubApi = require('../GitHubApi');
 const Promise = require('pinkie-promise');
 const renderer = require('../renderer');
+const translate = require('yandex-translate')(process.env.YANDEX_TRANSLATE_KEY);
```
Comment (AnyaTyu) [inline server/controllers/translator.js:6]: translate -> translater

File: server/controllers/translator.js [hunk 22]
```diff
@@ -57,7 +57,51 @@ function saveMemory(req, res) {
...
+        return new Promise((resolve, reject) => {
+            if (item.target) {
+                item.source = { content: item.source };
+                item.target = { content: item.target };
+                return resolve(item);
+            }
+            translate.translate(item.source, {
+                // TODO: here we should think for better solution
+                // maybe we should use simplified format anywhere
+                from: item.source_lang.slice(0,2),
```
Comment (AnyaTyu) [inline server/controllers/translator.js]: space needed: `0, 2`

File: server/controllers/translator.js [hunk 23]
```diff
@@ -57,7 +57,51 @@ function saveMemory(req, res) {
...
+                to:   item.target_lang.slice(0,2)
+            },
+            function(err, result) {
+                if (err) {
+                    console.error(err);
+                    reject({ code: 500, message: 'Something wrong happens' });
```
Comment (AnyaTyu) [inline server/controllers/translator.js]: return reject

File: server/controllers/translator.js [hunk 24]
```diff
@@ -57,7 +57,51 @@ function saveMemory(req, res) {
...
+                }
 
+                if (result.code === 200) {
+                    item.target = { content: result.text[0] };
+                    item.source = { content: item.source };
+                    return resolve(item);
+                } else {
+                    console.log(result.code, result.message);
```
Comment (AnyaTyu) [inline server/controllers/translator.js]: console.error

**Pull request closed without merging**